Jumpscale/clients/blockchain/tfchain/TFChainReservation.py

```diff
@@ -0,0 +1,105 @@
+from Jumpscale import j
+import nacl
+
+_RESERVATION_SCHEMA = j.data.schema.get("""
+@url = tfchain.reservation
+type = "S3,VM" (E)
+duration = (I)
+size = (I)
+email = (email)
+bot_id = (S)
+created = (I)
+""")
+
+_GRID_BROKER_PUB_KEY = "..."
+
+
+class TFChainReservation():
+    """
+    TFChainReservation contains all TF grid reservation logic.
+    """
+
+    def __init__(self, wallet):
+        self._wallet = wallet
+        self._notary_client = None
+        self._grid_broker_pub_key = None
+        self._grid_broker_addr = ''  # TODO: hardcode tfchain address of the grid broker
+        # this all key management needs to be improved
+        keypair = self._wallet.key_pair_get(self._wallet.address)
+        self._signing_key = nacl.signing.SigningKey(keypair.private_key.to_seed())
+        self._private_key = self._signing_key.to_curve25519_private_key()
+
+    @property
+    def notary_client(self):
+        if self._notary_client is None:
+            self._notary_client = j.clients.gedis.configure('tfnotary', host='notary.grid.tf', port=8888)
+        return self._notary_client
+
+    @property
+    def grid_broker_pub_key(self):
+        if self._grid_broker_pub_key is None:
+            record = self._wallet.client.threebot.record_get('tf3bot.zaibon')
+            encoded_key = record.public_key.unlockhash
+            self._grid_broker_pub_key = nacl.public.PublicKey(
+                str(encoded_key.hash),
+                encoder=nacl.encoding.HexEncoder)
+        return self._grid_broker_pub_key
+
+    def new(self):
```

Comment:
do we need to expose the schema as-is? It would be a much more user-friendly API if I can just reserve & submit in one call, as we do with the rest of the TFChain client. The goal of the API was to not require any non-primitive types. Or am I missing something here?

Reply:
I don't know yet how I'm going to do this.
I was thinking maybe having a type that encapsulate the raw schema object.

I kinda hate the one does all method cause this is just super hard not to break your interface when you do that.

Now at least I have a clear interface, `new`, `send`, if the reservation format change in the future, my interface stays the same, only the object return by new changed.
So any user of this lib can keep his code as is.

Reply:
I suppose, but it does make it more work to use, and goes against the API choices made in the rest of this API

Reply:
I see. It's a bit sad, but it is what it is I guess. I'll adapt the code

Reply:
Actually I think I'll revert my last change and go back to the `new`/`send` design.
Cause I can already know I'm going to have to break the API for phase2 of the launch.

Phase2 introduce 2 new workload to reserve, proxy and 0-db namespaces.
These will requires other field then what we have now.

So or I create a new reserve function per workload
or I go with my first idea, new/send.

I'm ok with both which one do you prefer @GlenDC ?

Reply:
How are you planning to introduce phase2 reservations? You'll require other fields. So will that mean you'll just add fields to your reservation struct, or are you planning to introduce a new reservation? Python has stuff like `**kwargs` which would allow for supporting both in 1 one function without breaking anyone. Need to think about this a bit, and play with the API usage on paper. Do you have a vague idea on how you would see phase 2 reservations, guesses are good enough, just so I have more stuff to thinker with and think about.

Reply:
Your approach `(1)`:

```python
reservation = j.clients.tfchain.capacity.reservation_new()
reservation.foo = foo
reservation.bar = bar
j.clients.tfchain.capacity.reservation_send(reservation)
```

> NOTE: I changed the namespace, I think reservations is a bit too vague, instead I
> would like to suggest to umbrella this logic under a "Capacity" wallet sub-client,
> as than it is immediately clear what the reservation is about.

Your suggestion is still easy enough to use. However, there is no good reason
why the variables desired for the properties cannot already be given to the
new constructor.

```python
reservation = j.clients.tfchain.capacity.reservation_new(foo=foo, bar=bar)
j.clients.tfchain.capacity.reservation_send(reservation)
```

If so, however, we can just as well imediately make it submit as well,
as there is no point in making the submit still another step, it would only
be another step the user has to learn without any added benefit.

```python
j.clients.tfchain.capacity.reservation_new(foo=foo, bar=bar) # submits as well
```

Further, let's say that you do use your approach `(1)` than:

- either you would add new fields to reservation struct, in which case you can just as well
  provide new `**kwargs` to a pure method-based approach;
- you would provide a new reservation struct, providing a new constructor approach,
  and allowing the different reservation structs to be passed to the send function,
  if so you can just as well provide several reseration_new functions,
  which I do suggest really if we talk about very different workloads
  (would need to have more context and info to decide on this)

Either way, it seems the choices that you have to make are very much in parellel for both
your approach and the approach the TFChain API has followed up till now.

Reply:
Let me first give more detail about what are the new reservation type in phase2.
The 2 new workload that will be available for reservation in phase2 are reverse proxy and 0-db namespace. 
Both of these required different field then what we currently have.

So my idea was to have a single factory that create a different reservation object depending on the type of workload you ask.
The kwargs can indeed works, but this is not user friendly. kwargs mask all the possible arguments. And since we expect people to discover the API using the shell, having explicit argument is always better.

That's what I think having something like:
```python
reservation = w.capacity.reservation_new(type=vm)
reservation.field1 = ...
reservation.field2 = ...
w.capacity.reservation_send(reservation)
```
Is still the best approach cause then user can just look at the field of the reservation object, fill them all in and then send it.
This is extendable to any number of reservation type.
We can also support `**kwargs` in the `reservation_new` method to pre-fill the reservation field if we want.

If we really don't want to have a 2 step process (new, send).
Then I think I need to have a different method per reservation type.
This is the more explicit we can do I think, since then I can have proper argument in each function.

Reply:
Problem with this approach is how will you let the user enter `vm`, using an Enum would be the logical choice but is very hard to do given the requirement that we want everything to be done via `j...`, and thus it would make using that function very long.

You could do it via a string, and cast it to a JS enum, but than you come to your described issue where the API is not discoverable via the shell (unless documented I guess, but still not as good as it can bet). Or do you think this is OK?

If not, as I think, perhaps it should just be a separate function for each reservation, e.g. `reservation_vm_new(...)`? But if so, then again we can just as well merge the 2 steps into a single step, as I do not see another reason left to keep intermediate object first.

Reply:
> If not, as I think, perhaps it should just be a separate function for each reservation, e.g. 

I think so too and if we go that way it can be a single function indeed.

Reply:
OK, than I guess we finally got the an agreement ^^